src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs

```diff
@@ -0,0 +1,265 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
+
+namespace System.Threading.RateLimiting
+{
+    /// <summary>
+    /// <see cref="RateLimiter"/> implementation that helps manage concurrent access to a resource.
+    /// </summary>
+    public sealed class ConcurrencyLimiter : RateLimiter
+    {
+        private int _permitCount;
+        private int _queueCount;
+
+        private readonly ConcurrencyLimiterOptions _options;
+        private readonly Deque<RequestRegistration> _queue = new Deque<RequestRegistration>();
+
+        private static readonly ConcurrencyLease SuccessfulLease = new ConcurrencyLease(true, null, 0);
+        private static readonly ConcurrencyLease FailedLease = new ConcurrencyLease(false, null, 0);
+        private static readonly ConcurrencyLease QueueLimitLease = new ConcurrencyLease(false, null, 0, "Queue limit reached");
+
+        // Use the queue as the lock field so we don't need to allocate another object for a lock and have another field in the object
+        private object Lock => _queue;
+
+        /// <summary>
+        /// Initializes the <see cref="ConcurrencyLimiter"/>.
+        /// </summary>
+        /// <param name="options">Options to specify the behavior of the <see cref="ConcurrencyLimiter"/>.</param>
+        public ConcurrencyLimiter(ConcurrencyLimiterOptions options)
+        {
+            _options = options;
+            _permitCount = _options.PermitLimit;
+        }
+
+        /// <inheritdoc/>
+        public override int GetAvailablePermits() => _permitCount;
+
+        /// <inheritdoc/>
+        protected override RateLimitLease AcquireCore(int permitCount)
+        {
+            // These amounts of resources can never be acquired
+            if (permitCount > _options.PermitLimit)
+            {
+                throw new ArgumentOutOfRangeException(nameof(permitCount), $"{permitCount} permits exceeds the permit limit of {_options.PermitLimit}.");
```

Comment:
Nit: could use the overload that accepts the value that's out of range.

```suggestion
                throw new ArgumentOutOfRangeException(nameof(permitCount), permitCount, $"{permitCount} permits exceeds the permit limit of {_options.PermitLimit}.");
```